instrumentation/kamon-zio-2/src/main/scala/kamon/instrumentation/zio2/ZIO2Instrumentation.scala

```diff
@@ -0,0 +1,60 @@
+package kamon.instrumentation.zio2
+
+import kamon.Kamon
+import kamon.instrumentation.context.HasContext
+import kanela.agent.api.instrumentation.InstrumentationBuilder
+import kanela.agent.libs.net.bytebuddy.asm.Advice
+import zio.{Exit, Fiber, Supervisor, UIO, Unsafe, ZEnvironment, ZIO}
+
+import scala.annotation.static
+
+/**
+ * This works as follows.
+ *  - patches the defaultSupervisor val from Runtime to add our own supervisor.
+ *  - Mixes in the [[HasContext.Mixin]] class so we don't have to keep a separate map of Fiber -> Context
+ *  - Performs context shifting based on starting/suspending of fibers.
+ *
+ */
+class ZIO2Instrumentation extends InstrumentationBuilder {
+
+  onType("zio.internal.FiberRuntime")
+    .mixin(classOf[HasContext.Mixin])
+    .advise(isConstructor, AfterFiberRuntimeInit)
+
+  onType("zio.Runtime$")
+    .advise(method("defaultSupervisor"), classOf[SupervisorAdvice.OverrideDefaultSupervisor])
+}
+
+class AfterFiberRuntimeInit
+
+object AfterFiberRuntimeInit  {
+
+  @Advice.OnMethodExit
+  @static def exit(@Advice.This fiber: Any): Unit ={
+    fiber.asInstanceOf[HasContext].setContext(Kamon.currentContext())
+  }
+
+}
+
+
+class NewSupervisor extends Supervisor[Any] {
+
+  override def value(implicit trace: zio.Trace): UIO[Any] = ZIO.unit
+
+  override def onStart[R, E, A_](environment: ZEnvironment[R], effect: ZIO[R, E, A_], parent: Option[Fiber.Runtime[Any, Any]], fiber: Fiber.Runtime[E, A_])(implicit unsafe: Unsafe): Unit = {
+    val ctx = fiber.asInstanceOf[HasContext].context
+    Kamon.storeContext(ctx)
+  }
+
+  override def onSuspend[E, A_](fiber: Fiber.Runtime[E, A_])(implicit unsafe: Unsafe): Unit = {
+    val pctx = Kamon.currentContext()
+    fiber.asInstanceOf[HasContext].setContext(pctx)
+  }
+
+  override def onResume[E, A_](fiber: Fiber.Runtime[E, A_])(implicit unsafe: Unsafe): Unit = {
+    val ctx = fiber.asInstanceOf[HasContext].context
+    Kamon.storeContext(ctx)
+  }
```

Comment:
Are these the right way around? That looks opposite from what I'd expect but IDK.

Reply:
on suspend you're setting the context of the fiber that you had back to the fiber itself where you were executing it. 

onResume your taking the fiber from the pool that you are resuming and applying context to the current session.  I am probably saying that wrong.